bundles/org.openhab.binding.teslapowerwall/src/main/java/org/openhab/binding/teslapowerwall/internal/TeslaPowerwallWebTargets.java

```diff
@@ -0,0 +1,133 @@
+        String baseUri = "https://" + ipAddress + "/";
+        getBatterySOEUri = baseUri + "api/system_status/soe";
+        getGridStatusUri = baseUri + "api/system_status/grid_status";
+        getSystemStatusUri = baseUri + "api/system_status";
+        getMeterAggregatesUri = baseUri + "api/meters/aggregates";
+        getTokenUri = baseUri + "api/login/Basic";
+        getOperationUri = baseUri + "api/operation";
+    }
+
+    public BatterySOE getBatterySOE(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getBatterySOEUri, token);
+        logger.trace("getBatterySOE response = {}", response);
+        return BatterySOE.parse(response);
+    }
+
+    public GridStatus getGridStatus(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getGridStatusUri, token);
+        logger.trace("getGridStatus response = {}", response);
+        return GridStatus.parse(response);
+    }
+
+    public SystemStatus getSystemStatus(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getSystemStatusUri, token);
+        logger.trace("getSystemStatus response = {}", response);
+        return SystemStatus.parse(response);
+    }
+
+    public MeterAggregates getMeterAggregates(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getMeterAggregatesUri, token);
+        logger.trace("getMeterAggregates response = {}", response);
+        return MeterAggregates.parse(response);
+    }
+
+    public Operations getOperations(String token) throws TeslaPowerwallCommunicationException {
+        String response = invoke(getOperationUri, token);
+        logger.trace("getOperations response = {}", response);
+        return Operations.parse(response);
+    }
+
+    public String getToken(String email, String password) throws TeslaPowerwallCommunicationException {
+        JsonObject jsonObject = new JsonObject();
+        jsonObject.addProperty("username", "customer");
+        jsonObject.addProperty("password", password);
+        jsonObject.addProperty("email", email);
+        jsonObject.addProperty("force_sm_off", false);
+        logger.debug("logonjson = {}", jsonObject.toString());
+        String response = invoke(getTokenUri, "POST", "application/json", jsonObject.toString());
+        JsonObject jsonResponse = JsonParser.parseString(response).getAsJsonObject();
+        String token = jsonResponse.get("token").getAsString();
+        logger.debug("Token: {}", token);
+        return token;
+    }
+
+    private String invoke(String uri, String token) throws TeslaPowerwallCommunicationException {
+        return invoke(uri, "GET", token, "");
+    }
+
+    private String invoke(String uri, String request, String contenttype, String params)
+            throws TeslaPowerwallCommunicationException {
+        logger.debug("Calling url: {}", uri);
```

Comment:
This method does nothing with the token, but i guess this is the best place to check if the token is set, valid or expired, renew and continue the actuall call. 
This way this class is fully responsible for communcation with the remote service including token handling. The caller class `TeslaPowerwallHandler.java` does not need to know anything about the token.
I would recommend to add the username and password to the constructor, just like the ipaddress.

This would improve the seperation of concern.

Reply:
I can see the value in this,  but it will be quite intrusive,  so I need to wake up a bit more before I tackle this!

Reply:
Take your time, no need to rush

Reply:
I want to get this binding into OH4.2.0 so I'll aim to complete this over the weekend. It's  a holiday weekend for us in Australia :)

Reply:
I'm sure we manage to merg this before the 29th of june (last milestone / feature freeze)

Reply:
OK I've now implemented this.